File: weird_captcha_gym/shared_runtime/server/incubator_graders/three_trade_crew.py
from __future__ import annotations

import copy
from typing import Any


MECHANIC_ID = "three_trade_crew"
DIRECTIONS = {"up": (-1, 0), "down": (1, 0), "left": (0, -1), "right": (0, 1)}


def _fail(feedback: str) -> dict[str, Any]:
    return {"graded": True, "passed": False, "feedback": feedback}


def _bind(payload: dict[str, Any], truth: dict[str, Any], public: dict[str, Any]) -> str | None:
    if payload.get("mechanic_id") != MECHANIC_ID or truth.get("mechanic_id") != MECHANIC_ID or public.get("mechanic_id") != MECHANIC_ID:
        return "mechanic binding failed"
    challenge = str(truth.get("challenge_id") or "")
    task_id = str(truth.get("task_id") or "")
    if not challenge or payload.get("challenge_id") != challenge or public.get("challenge_id") != challenge:
        return "stale challenge"
    if not task_id or payload.get("task_id") != task_id or public.get("task_id") != task_id:
        return "task binding failed"
    if truth.get("control_condition") != public.get("control_condition"):
        return "control condition mismatch"
    return None


def _point(value: Any) -> tuple[int, int]:
    if not isinstance(value, list) or len(value) != 2 or any(isinstance(item, bool) or not isinstance(item, int) for item in value):
        raise ValueError("position must be an integer pair")
    return int(value[0]), int(value[1])


def _board(value: Any) -> dict[str, Any]:
    if not isinstance(value, dict):
        raise ValueError("board is missing")
    result = copy.deepcopy(value)
    for key in ("rows", "columns"):
        if isinstance(result.get(key), bool) or not isinstance(result.get(key), int):
            raise ValueError(f"board {key} is invalid")
    for key in ("walls",):
        if not isinstance(result.get(key), list):
            raise ValueError(f"board {key} is invalid")
    for item in result["walls"]:
        _point(item)
    for key in ("workers", "crates", "goals", "switches", "doors"):
        if not isinstance(result.get(key), list):
            raise ValueError(f"board {key} is invalid")
    return result


def _occupied(board: dict[str, Any]) -> dict[tuple[int, int], tuple[str, str]]:
    result: dict[tuple[int, int], tuple[str, str]] = {}
    for worker in board["workers"]:
        result[_point(worker["position"])] = ("worker", str(worker["id"]))
    for crate in board["crates"]:
        result[_point(crate["position"])] = ("crate", str(crate["id"]))
    return result


def _switch_open(board: dict[str, Any], switch_id: str) -> bool:
    switch = next(item for item in board["switches"] if item["id"] == switch_id)
    point = _point(switch["position"])
    return any(_point(item["position"]) == point for item in board["workers"] + board["crates"])


def _passable(board: dict[str, Any], point: tuple[int, int], worker_id: str) -> bool:
    row, column = point
    if not 0 <= row < board["rows"] or not 0 <= column < board["columns"]:
        return False
    if point in {_point(item) for item in board["walls"]}:
        return False
    for door in board["doors"]:
        if _point(door["position"]) == point and not _switch_open(board, str(door["switch_id"])):
            return False
    occupant = _occupied(board).get(point)
    return occupant is None or occupant[0] == "crate" or occupant == ("worker", worker_id)


def _apply(board: dict[str, Any], worker_id: str, direction: str) -> tuple[dict[str, Any], dict[str, Any]]:
    if direction not in DIRECTIONS:
        raise ValueError("unknown direction")
    current = copy.deepcopy(board)
    worker = next((item for item in current["workers"] if item["id"] == worker_id), None)
    if worker is None:
        raise ValueError("unknown worker")
    dr, dc = DIRECTIONS[direction]
    old = _point(worker["position"])
    dest = (old[0] + dr, old[1] + dc)
    if not _passable(current, dest, worker_id):
        raise ValueError("blocked by wall or closed door")
    if _occupied(current).get(dest, (None, None))[0] == "worker":
        raise ValueError("workers cannot overlap")
    crate = next((item for item in current["crates"] if _point(item["position"]) == dest), None)
    ability = "move"
    crate_id = None
    if crate is not None:
        crate_id = str(crate["id"])
        if worker["role"] == "fighter":
            beyond = (dest[0] + dr, dest[1] + dc)
            if not _passable(current, beyond, worker_id) or beyond in _occupied(current):
                raise ValueError("fighter cannot push")
            crate["position"] = [*beyond]
            ability = "push"
        elif worker["role"] in {"thief", "wizard"}:
            crate["position"] = [*old]
            ability = "pull" if worker["role"] == "thief" else "swap"
    elif worker["role"] == "thief":
        behind = (old[0] - dr, old[1] - dc)
        trailing = next((item for item in current["crates"] if _point(item["position"]) == behind), None)
        if trailing is not None:
            trailing["position"] = [*old]
            crate_id = str(trailing["id"])
            ability = "pull"
    worker["position"] = [*dest]
    return current, {
        "worker_id": worker_id,
        "direction": direction,
        "from": [*old],
        "to": [*dest],
        "ability": ability,
        "crate_id": crate_id,
        "door_states": {str(door["id"]): _switch_open(current, str(door["switch_id"])) for door in current["doors"]},
        "workers_after": copy.deepcopy(current["workers"]),
        "crates_after": copy.deepcopy(current["crates"]),
    }


def _solved(board: dict[str, Any]) -> bool:
    return all(_point(worker["position"]) == _point(next(goal["position"] for goal in board["goals"] if goal["worker_id"] == worker["id"])) for worker in board["workers"])


def _contract(truth: dict[str, Any], public: dict[str, Any]) -> tuple[dict[str, Any], dict[str, Any], str]:
    parameters = truth.get("parameters")
    if not isinstance(parameters, dict) or parameters != public.get("parameters"):
        raise ValueError("difficulty parameters differ")
    hidden = _board(truth.get("initial_board"))
    visible = _board((public.get("board") or {}))
    if hidden != visible:
        raise ValueError("visible board differs from generated board")
    condition = truth.get("control_condition") or {}
    interaction = str(condition.get("interaction") or "")
    if interaction not in {"full", "simplified"}:
        raise ValueError("invalid interaction condition")
    return hidden, parameters, interaction


def grade(payload: dict[str, Any], truth: dict[str, Any], public: dict[str, Any]) -> dict[str, Any]:
    binding = _bind(payload, truth, public)
    if binding:
        return _fail(binding)
    try:
        board, parameters, interaction = _contract(truth, public)
    except (TypeError, ValueError) as exc:
        return _fail(f"invalid crew contract: {exc}")
    if payload.get("interaction_mode") != interaction:
        return _fail("submitted interaction surface differs from the selected condition")
    expected_action_source = "keyboard_move" if interaction == "full" else "direction_button"
    expected_select_source = "keyboard_select" if interaction == "full" else "worker_card"
    selections = payload.get("selection_events")
    events = payload.get("events")
    if not isinstance(selections, list) or not isinstance(events, list):
        return _fail("selection or movement transcript is missing")
    if len(events) > int(parameters.get("action_budget", 999)):
        return _fail("movement transcript exceeds the visible action budget")
    workers = {str(item["id"]): item for item in board["workers"]}
    selected: str | None = None
    selection_index = 0
    ordered_selections = sorted(selections, key=lambda item: int(item.get("movement_index", 0)) if isinstance(item, dict) else 0)
    selection_cursor = 0
    for selection in ordered_selections:
        if not isinstance(selection, dict) or selection.get("input_source") != expected_select_source:
            return _fail("selection uses the wrong interaction surface")
        worker_id = str(selection.get("worker_id") or "")
        if worker_id not in workers:
            return _fail("selection names an unavailable worker")
        selection_index += 1
    used = set()
    for index, event in enumerate(events, start=1):
        if not isinstance(event, dict) or event.get("sequence") != index:
            return _fail(f"movement {index} has an invalid sequence")
        if event.get("input_source") != expected_action_source:
            return _fail(f"movement {index} uses the wrong interaction surface")
        while selection_cursor < len(ordered_selections) and int(ordered_selections[selection_cursor].get("movement_index", 0)) <= index - 1:
            selected = str(ordered_selections[selection_cursor].get("worker_id") or "")
            selection_cursor += 1
        worker_id = str(event.get("worker_id") or "")
        if selected != worker_id:
            return _fail(f"movement {index} was sent to a worker that was not selected")
        before = next(item for item in board["workers"] if item["id"] == worker_id)
        if event.get("from") != before["position"]:
            return _fail(f"movement {index} reports the wrong origin")
        try:
            board, expected = _apply(board, worker_id, str(event.get("direction") or ""))
        except (TypeError, ValueError) as exc:
            return _fail(f"movement {index} is illegal: {exc}")
        for field in ("to", "ability", "crate_id", "door_states", "workers_after", "crates_after"):
            if event.get(field) != expected[field]:
                return _fail(f"movement {index} has inconsistent {field}")
        used.add(expected["ability"])
    final = payload.get("final_board")
    if not isinstance(final, dict) or final.get("workers") != board["workers"] or final.get("crates") != board["crates"]:
        return _fail("final workshop state does not match movement replay")
    completed = payload.get("completed") is True
    passed = completed and _solved(board) and {"push", "pull", "swap"}.issubset(used)
    if completed and not passed:
        return _fail("certification was claimed before all three workers reached their bays")
    return {"graded": True, "passed": passed, "feedback": f"replayed {len(events)} movements, selected {selection_index} times, abilities={','.join(sorted(used))}"}


def cheat(public_state: dict[str, Any], ground_truth: dict[str, Any]) -> dict[str, Any]:
    return {"solution_actions": ground_truth.get("solution_actions") or [], "instruction": "Select each worker and perform the visible crew movements in order.", "answers": []}

File: weird_captcha_gym/shared_runtime/app/mechanics/three_trade_crew.js
(() => {
  "use strict";

  let model = null;
  const clone = (value) => JSON.parse(JSON.stringify(value));
  const esc = (value) => String(value == null ? "" : value).replaceAll("&", "&amp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;").replaceAll('"', "&quot;");
  const deltas = {up: [-1, 0], down: [1, 0], left: [0, -1], right: [0, 1]};
  const directionLabel = {up: "NORTH", down: "SOUTH", left: "WEST", right: "EAST"};
  const interaction = () => model?.state?.control_condition?.interaction || "full";
  const workerById = (id) => model.board.workers.find((worker) => worker.id === id);

  function occupied() {
    const result = new Map();
    model.board.workers.forEach((worker) => result.set(worker.position.join(","), {kind: "worker", id: worker.id}));
    model.board.crates.forEach((crate) => result.set(crate.position.join(","), {kind: "crate", id: crate.id}));
    return result;
  }

  function switchOpen(switchId) {
    const sw = model.board.switches.find((item) => item.id === switchId);
    if (!sw) return false;
    const point = sw.position.join(",");
    return model.board.workers.concat(model.board.crates).some((item) => item.position.join(",") === point);
  }

  function passable(point, workerId) {
    const [row, column] = point;
    if (row < 0 || row >= model.board.rows || column < 0 || column >= model.board.columns) return false;
    if (model.board.walls.some((wall) => wall[0] === row && wall[1] === column)) return false;
    const door = model.board.doors.find((item) => item.position[0] === row && item.position[1] === column);
    if (door && !switchOpen(door.switch_id)) return false;
    const occupant = occupied().get(`${row},${column}`);
    return !occupant || occupant.kind === "crate" || (occupant.kind === "worker" && occupant.id === workerId);
  }

  function applyMove(workerId, direction) {
    const next = clone(model.board);
    const worker = next.workers.find((item) => item.id === workerId);
    if (!worker || !deltas[direction]) return null;
    const [dr, dc] = deltas[direction];
    const old = [...worker.position];
    const dest = [old[0] + dr, old[1] + dc];
    const oldBoard = model.board;
    const oldOccupied = occupied();
    if (!passable(dest, workerId)) return null;
    const occupant = oldOccupied.get(dest.join(","));
    if (occupant?.kind === "worker") return null;
    const crate = next.crates.find((item) => item.position[0] === dest[0] && item.position[1] === dest[1]);
    let ability = "move";
    let crateId = null;
    if (crate) {
      crateId = crate.id;
      if (worker.role === "fighter") {
        const beyond = [dest[0] + dr, dest[1] + dc];
        const beyondOccupied = oldOccupied.get(beyond.join(","));
        if (!passable(beyond, workerId) || beyondOccupied) return null;
        crate.position = beyond;
        ability = "push";
      } else {
        crate.position = old;
        ability = worker.role === "thief" ? "pull" : "swap";
      }
    } else if (worker.role === "thief") {
      const behind = [old[0] - dr, old[1] - dc];
      const trailing = next.crates.find((item) => item.position[0] === behind[0] && item.position[1] === behind[1]);
      if (trailing) {
        trailing.position = old;
        crateId = trailing.id;
        ability = "pull";
      }
    }
    worker.position = dest;
    const doorStates = {};
    next.doors.forEach((door) => {
      const sw = next.switches.find((item) => item.id === door.switch_id);
      const point = sw?.position?.join(",");
      doorStates[door.id] = next.workers.concat(next.crates).some((item) => item.position.join(",") === point);
    });
    return {
      board: next,
      info: {
        worker_id: workerId,
        direction,
        from: old,
        to: dest,
        ability,
        crate_id: crateId,
        door_states: doorStates,
        workers_after: clone(next.workers),
        crates_after: clone(next.crates),
      },
    };
  }

  function solved() {
    return model.board.workers.every((worker) => {
      const goal = model.board.goals.find((item) => item.worker_id === worker.id);
      return goal && goal.position[0] === worker.position[0] && goal.position[1] === worker.position[1];
    });
  }

  function selectWorker(workerId, inputSource) {
    if (model.busy || model.passed || !workerById(workerId)) return;
    document.querySelector(".ttc-verdict-layer .ttc-fail")?.remove();
    model.selected = workerId;
    model.selections.push({worker_id: workerId, input_source: inputSource, movement_index: model.events.length});
    update();
    model.helpers.setReadout(`${workerById(workerId).label} ACTIVE`, "idle");
  }

  function move(direction, inputSource) {
    if (model.busy || model.passed || !model.selected) return;
    document.querySelector(".ttc-verdict-layer .ttc-fail")?.remove();
    const result = applyMove(model.selected, direction);
    if (!result) {
      model.helpers.setReadout("BLOCKED", "error");
      return;
    }
    model.board = result.board;
    const event = result.info;
    event.sequence = model.events.length + 1;
    event.input_source = inputSource;
    model.events.push(event);
    update();
    model.helpers.setReadout(solved() ? "ALL BAYS READY" : `${event.ability.toUpperCase()} · ${directionLabel[direction]}`, solved() ? "passed" : "idle");
  }

  function workerMarkup(worker) {
    const selected = worker.id === model.selected ? " is-selected" : "";
    return `<button type="button" class="ttc-worker-card${selected}" data-worker-id="${esc(worker.id)}" aria-label="${esc(worker.label)}">
      <span class="ttc-worker-icon role-${esc(worker.role)}"><i></i><b></b></span><span><strong>${esc(worker.label)}</strong><small>${esc(worker.role === "fighter" ? "PUSH" : worker.role === "thief" ? "PULL" : "SWAP")}</small></span>
    </button>`;
  }

  function cellMarkup(row, column) {
    const key = `${row},${column}`;
    const wall = model.board.walls.some((item) => item[0] === row && item[1] === column);
    if (wall) return `<div class="ttc-cell ttc-wall" data-row="${row}" data-column="${column}"><span></span></div>`;
    const door = model.board.doors.find((item) => item.position[0] === row && item.position[1] === column);
    const sw = model.board.switches.find((item) => item.position[0] === row && item.position[1] === column);
    const goal = model.board.goals.find((item) => item.position[0] === row && item.position[1] === column);
    const worker = model.board.workers.find((item) => item.position[0] === row && item.position[1] === column);
    const crate = model.board.crates.find((item) => item.position[0] === row && item.position[1] === column);
    const classes = ["ttc-cell"];
    if (door) classes.push("ttc-door", `door-${door.color}`, switchOpen(door.switch_id) ? "is-open" : "is-closed");
    if (sw) classes.push("ttc-switch", `switch-${sw.color}`);
    if (goal) classes.push("ttc-goal");
    let content = "";
    if (goal) content += `<span class="ttc-goal-mark">${esc(goal.label)}</span>`;
    if (sw) content += `<span class="ttc-switch-mark">${esc(sw.label)}</span>`;
    if (door) content += `<span class="ttc-door-mark">${switchOpen(door.switch_id) ? "OPEN" : "LOCK"}</span>`;
    if (crate) content += `<span class="ttc-crate" style="--crate-color:${esc(crate.color)}"><i></i><b>${esc(crate.id.slice(-1))}</b></span>`;
    if (worker) content += `<span class="ttc-worker-token role-${esc(worker.role)}${worker.id === model.selected ? " is-selected" : ""}"><i></i><b>${esc(worker.label[0])}</b></span>`;
    return `<div class="${classes.join(" ")}" data-row="${row}" data-column="${column}">${content}</div>`;
  }

  function boardMarkup() {
    const cells = [];
    for (let row = 0; row < model.board.rows; row += 1) {
      for (let column = 0; column < model.board.columns; column += 1) cells.push(cellMarkup(row, column));
    }
    return cells.join("");
  }

  function controlsMarkup() {
    const dirs = ["up", "left", "down", "right"];
    return `<div class="ttc-direction-pad">${dirs.map((dir) => `<button type="button" data-direction="${dir}"><span>${dir === "up" ? "▲" : dir === "down" ? "▼" : dir === "left" ? "◀" : "▶"}</span><small>${directionLabel[dir]}</small></button>`).join("")}</div>`;
  }

  function update() {
    const board = document.querySelector(".ttc-board");
    if (board) board.innerHTML = boardMarkup();
    document.querySelectorAll(".ttc-worker-card").forEach((button) => button.classList.toggle("is-selected", button.dataset.workerId === model.selected));
    document.querySelector(".ttc-count")?.replaceChildren(document.createTextNode(`${model.events.length} MOVES`));
    bindControls();
  }

  function bindControls() {
    document.querySelectorAll(".ttc-worker-card").forEach((button) => {
      if (interaction() === "simplified") button.onclick = () => selectWorker(button.dataset.workerId, "worker_card");
    });
    document.querySelectorAll(".ttc-direction-pad button").forEach((button) => {
      button.onclick = () => move(button.dataset.direction, "direction_button");
    });
  }

  function keyHandler(event) {
    if (interaction() !== "full" || event.repeat || model.busy || model.passed) return;
    const selectMap = {Digit1: "fighter", Digit2: "thief", Digit3: "wizard"};
    if (selectMap[event.code]) { event.preventDefault(); selectWorker(selectMap[event.code], "keyboard_select"); return; }
    if (event.key.toLowerCase() === "x") {
      event.preventDefault();
      const index = model.board.workers.findIndex((worker) => worker.id === model.selected);
      selectWorker(model.board.workers[(index + 1 + model.board.workers.length) % model.board.workers.length].id, "keyboard_select");
      return;
    }
    const direction = {ArrowUp: "up", ArrowDown: "down", ArrowLeft: "left", ArrowRight: "right"}[event.key];
    if (direction) { event.preventDefault(); move(direction, "keyboard_move"); }
  }

  async function submit(completed) {
    if (model.busy || model.passed) return;
    model.busy = true;
    model.helpers.setReadout("CHECKING", "pending");
    const payload = {
      mechanic_id: model.state.mechanic_id,
      task_id: model.state.task_id,
      challenge_id: model.state.challenge_id,
      interaction_mode: interaction(),
      selection_events: clone(model.selections),
      events: clone(model.events),
      final_board: {workers: clone(model.board.workers), crates: clone(model.board.crates)},
      completed: completed === true,
    };
    try {
      const response = await fetch("/result", {method: "POST", headers: {"content-type": "application/json"}, body: JSON.stringify(payload)});
      const outcome = await response.json();
      if (outcome.passed === true) {
        model.busy = false; model.passed = true;
        document.querySelector(".three-trade-crew")?.classList.add("is-pass");
        document.querySelector(".ttc-verdict-layer .ttc-fail")?.remove();
        document.querySelector(".ttc-verdict-layer")?.insertAdjacentHTML("beforeend", '<div class="ttc-verdict ttc-pass"><strong>PASS</strong><small>CREW PARKED</small></div>');
        model.helpers.setReadout("PASS", "passed");
      } else if (outcome.passed === false && outcome.state) {
        await render(outcome.state, model.helpers, {freshFailure: true});
        model.helpers.setReadout("FAIL", "error");
      } else {
        model.busy = false; model.helpers.setReadout("FAIL", "error");
      }
    } catch (_error) {
      model.busy = false; model.helpers.setReadout("FAIL", "error");
    }
  }

  async function render(state, helpers, options = {}) {
    document.body.dataset.mechanic = "three-trade-crew";
    model = {state, helpers, board: clone(state.board), selected: "fighter", selections: [], events: [], busy: false, passed: false};
    helpers.app.innerHTML = `<section class="three-trade-crew mode-${esc(interaction())}" data-challenge-id="${esc(state.challenge_id)}">
      <header class="ttc-header"><div class="ttc-badge"><i></i><b>3</b></div><div><small>THE THREE-TRADE WORKSHOP</small><h1>${esc(state.prompt)}</h1></div><div class="ttc-count">0 MOVES</div></header>
      <main class="ttc-main"><aside class="ttc-sidebar"><div class="ttc-panel-title">CREW ROSTER</div><div class="ttc-roster">${model.board.workers.map(workerMarkup).join("")}</div></aside><section class="ttc-arena"><div class="ttc-arena-label"><span>WORKSHOP FLOOR</span><i></i><span>EXIT BAYS F / T / W</span></div><div class="ttc-board" style="--ttc-cols:${state.board.columns};--ttc-rows:${state.board.rows}">${boardMarkup()}</div></section><aside class="ttc-controls"><div class="ttc-panel-title">${interaction() === "full" ? "KEYBOARD DECK" : "CONTROL DECK"}</div>${interaction() === "full" ? '<p class="ttc-key-help">1 / 2 / 3 selects a worker · X cycles · ARROW KEYS move</p>' : '<p class="ttc-key-help">Select a worker, then send one square at a time.</p>'}${controlsMarkup()}</aside></main>
      <footer class="ttc-footer"><button type="button" id="ttc-reset">RESET VIEW</button><div class="ttc-status"><div class="readout" data-status="idle">${options.freshFailure ? "FAIL · FRESH WORKSHOP" : ""}</div></div><button type="button" id="ttc-certify">CERTIFY CREW</button></footer><div class="ttc-verdict-layer">${options.freshFailure ? '<div class="ttc-verdict ttc-fail"><strong>FAIL</strong><small>FRESH WORKSHOP LOADED</small></div>' : ""}</div>${helpers.cheatPanelTemplate()}</section>`;
    bindControls();
    document.getElementById("ttc-certify")?.addEventListener("click", () => submit(solved()));
    document.getElementById("ttc-reset")?.addEventListener("click", () => { model.board = clone(model.state.board); model.selections = []; model.events = []; model.selected = "fighter"; update(); model.helpers.setReadout("RESET", "idle"); });
    window.removeEventListener("keydown", keyHandler);
    window.addEventListener("keydown", keyHandler);
    helpers.installCheatPanel();
    window.threeTradeCrewModel = model;
  }

  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};
  window.WeirdCaptchaMechanics.three_trade_crew = {rootSelector: ".three-trade-crew", render};
})();
